lib/os_io.c

```diff
@@ -37,6 +37,199 @@ ___io_module ___io_mod =
+  memset (exceptfds, 0, bytes);
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = size;
+
+  return 0;
+
+ error:
+  if (readfds)
+    ___FREE_MEM (readfds);
+  if (writefds)
+    ___FREE_MEM (writefds);
+  if (exceptfds)
+    ___FREE_MEM (exceptfds);
+
+  return 1;
+}
+
+___HIDDEN int ___fdset_realloc
+   ___P((___processor_state ps,
+         int fd),
+        (ps,
+         fd)
+        ____processor_state ps;
+        int fd;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+  int oldbytes, newbytes;
+  int oldsize = ps->os.fdset.size;
+  int newsize = oldsize;
+
+  while (newsize <= fd)
+    newsize = newsize * 2;
+
+  if (oldsize == newsize) /* size unchanged, no need to realloc */
+      return 0;
+
+  oldbytes = ___CEILING_DIV (oldsize,8);
+  newbytes = ___CEILING_DIV (newsize,8);
+
+  readfds  = ___ALLOC_MEM (newbytes);
+  if (!readfds)
+    goto error;
+
+  writefds = ___ALLOC_MEM (newbytes);
+  if (!writefds)
+    goto error;
+
+#ifdef USE_select
+  exceptfds = ___ALLOC_MEM (newbytes);
+  if (!exceptfds)
+    goto error;
+#endif
+
+  memcpy (readfds, ps->os.fdset.readfds, oldbytes);
+  memset (___CAST(char*, readfds) + oldbytes, 0, newbytes - oldbytes);
+  memcpy (writefds, ps->os.fdset.writefds, oldbytes);
+  memset (___CAST(char*, writefds) + oldbytes, 0, newbytes - oldbytes);
+#ifdef USE_select
+  memcpy (exceptfds, ps->os.fdset.exceptfds, oldbytes);
+  memset (___CAST(char*, exceptfds) + oldbytes, 0, newbytes - oldbytes);
+#endif
+
+  ___FREE_MEM (ps->os.fdset.readfds);
+  ___FREE_MEM (ps->os.fdset.writefds);
+#ifdef USE_select
+  ___FREE_MEM (ps->os.fdset.exceptfds);
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = newsize;
+
+  return 0;
+
+ error:
+  if (readfds)
+    ___FREE_MEM (readfds);
+  if (writefds)
+    ___FREE_MEM (writefds);
+  if (exceptfds)
+    ___FREE_MEM (exceptfds);
+
+  return 1;
+}
+
+___HIDDEN int ___fdset_size
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ps->os.fdset.size;
+}
+
+___HIDDEN ___fdbits *___fdset_readfds
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ___CAST(___fdbits*, ps->os.fdset.readfds);
+}
+
+___HIDDEN ___fdbits *___fdset_writefds
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ___CAST(___fdbits*, ps->os.fdset.writefds);
+}
+
+___HIDDEN ___fdbits *___fdset_exceptfds
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  return ___CAST(___fdbits*, ps->os.fdset.exceptfds);
+}
+
+void ___fdset_resize_heap_overflow_clear ___PVOID
+{
+  ___fdset_heap_overflow = 0;
+}
+
+int ___fdset_resize_heap_overflow ___PVOID
+{
+  return ___fdset_heap_overflow;
+}
+
+#endif
+
+void ___fdset_resize_pstate
```

Comment:
This will require an ifdef guard in `execute_sync_op` as well, as it's the callsite for this function.